stan/math/opencl/normal_id_glm_lpdf.hpp

```diff
@@ -0,0 +1,199 @@
+#ifndef STAN_MATH_OPENCL_NORMAL_ID_GLM_LPDF_HPP
+#define STAN_MATH_OPENCL_NORMAL_ID_GLM_LPDF_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/scal/err/check_consistent_sizes.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
+#include <stan/math/prim/scal/fun/constants.hpp>
+#include <stan/math/prim/mat/fun/value_of_rec.hpp>
+#include <stan/math/prim/scal/fun/size_zero.hpp>
+#include <stan/math/prim/scal/fun/sum.hpp>
+#include <stan/math/prim/arr/fun/value_of_rec.hpp>
+
+#include <stan/math/opencl/kernels/normal_id_glm_lpdf.hpp>
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/multiply.hpp>
+
+#include <cmath>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the log PDF of the Generalized Linear Model (GLM)
+ * with Normal distribution and id link function.
+ * If containers are supplied, returns the log sum of the probabilities.
+ * The idea is that normal_id_glm_lpdf(y, x, alpha, beta, sigma) should
+ * compute a more efficient version of normal_lpdf(y, alpha + x * beta, sigma)
+ * by using analytically simplified gradients.
+ * @tparam T_alpha type of the intercept(s);
+ * this can be a vector (of the same length as y) of intercepts or a single
+ * value (for models with constant intercept);
+ * @tparam T_beta type of the weight vector;
+ * this can also be a single value;
+ * @tparam T_scale type of the (positive) scale(s);
+ * this can be a vector (of the same length as y, for heteroskedasticity)
+ * or a scalar.
+ * @param y_cl vector parameter on OpenCL device
+ * @param x_cl design matrix on OpenCL device
+ * @param alpha intercept (in log odds)
+ * @param beta weight vector
+ * @param sigma (Sequence of) scale parameters for the normal
+ * distribution.
+ * @return log probability or log sum of probabilities
+ * @throw std::domain_error if x, beta or alpha is infinite.
+ * @throw std::domain_error if the scale is not positive.
+ * @throw std::invalid_argument if container sizes mismatch.
+ */
+template <bool propto, typename T_alpha, typename T_beta, typename T_scale>
+return_type_t<T_alpha, T_beta, T_scale> normal_id_glm_lpdf(
+    const matrix_cl<double> &y_cl, const matrix_cl<double> &x_cl,
```

Comment:
This looks good to me from the stanc3 perspective! We'll need to specify in the type signatures list which arguments are required to be opencl if the fn is to execute on opencl so we know which ones to promote.